utils/FileHandler.cpp

```diff
@@ -0,0 +1,291 @@
+    qin.readLine();
+    qin.readLine();
+
+    return shape;
+}
+
+Shape* FileHandler::extractNonText(QTextStream& qin, QString& shapeType, QList<int>& dimensionsList)
+{
+    Shape* shape;    // Shape pointer to be returned by the function
+
+    // Extract data for the pen that draws this shape' edges
+    extractPen(qin);
+
+    // If the shape being read in isn't a line, we know it needs a brush
+    if(shapeType != jconstants::LINE && shapeType != jconstants::POLYLINE)
+    {
+        // Extract data for the brush used to fill the insides of this shape
+        extractBrush(qin);
+        shape = constructShape(shapeType, dimensionsList);
+    }
+    else
+    {
+        shape = constructLine(shapeType, dimensionsList);
+    }
+
+    return shape;
+}
+
+Shape *FileHandler::extractText(QTextStream &qin, QList<int> &dimensionsList)
+{
+    QString body;   // Body of the text
+    QFont font; // Font of the text
+
+    // Get to the next whitspace, then read in the line as the text's body
+    jbrush::getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    body = qin.readLine();
+
+    // For now, ignore the shape color and shape alignment
+    qin.readLine();
+    qin.readLine();
+
+    // Extract the text's font
+    font = extractFont(qin);
+
+    return new Text(QPoint(dimensionsList[0], dimensionsList[1]), body, font);
+}
+
+QList<int> FileHandler::extractDimensions(QTextStream &qin, const QString& shapeType)
+{
+    QList<int> dimensionsList;
+    QChar buffer;   // Used to track where the text stream is in the file
+    int dimension;  // Used to store the current dimension being read in
+    int expectedDimensionsSize; // The expected number of
+
+    // Get to next whitespace after leading title
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+
+    // Continue reading in integers until one is found that is not followed by the designated dimension interruptor
+    do
+    {
+        qin >> dimension;
+        dimensionsList.append(dimension);
+        qin >> buffer;
+    }while(buffer == jconstants::DIMENSION_INTERRUPTOR && !qin.atEnd());
+
+    // Check to make sure the expected number of dimensions were extracted
+    if(shapeType != jconstants::POLYLINE && shapeType != jconstants::POLYGON)
+    {
+        expectedDimensionsSize = jconstants::DIMENSION_COUNT_MAP[shapeType];
+
+        if(dimensionsList.size() != expectedDimensionsSize)
+        {
+            QString message;
+            QTextStream stream(&message);
+            stream << "The extractDimensions methdod in the FileHandler got " << dimensionsList.size() << " dimensions for the " << shapeType << "; expected " << expectedDimensionsSize;
+            throw BadFileFormat(message);
+        }
+    }
+    else if(dimensionsList.count() < 4)
+    {
+        QString message;
+        QTextStream stream(&message);
+        stream << "From extractDimensions method in FileHandler: a " << shapeType << " cannot be created from only " << dimensionsList.count() << " dimensions";
```

Comment:
Your throws exception messages are inconsistent from one another. That's why I suggested just being able to pass in the necessary things for the error message. For this particular error message, the throw should look like
```
throw BadFileFormat("extractDimensions", "more than 3 numbers", std::to_string(dimensionsList.count()) +"  
 numbers");
```
Which results in
```
extractDimensions: Expected more than 3 numbers but received 2 numbers.
```

And if an error message can't be simply put in this, just throw a ``GeneralException`` if you really have to.